Bug #130: Support for JSON-RPC-like Debug Server Protocol's RPC

jonahkichwacoders opened the pull request:
This provides the supporting RPC implementation for VSCode's Debug Protocol
as defined https://github.com/Microsoft/vscode-debugadapter-node/blob/master/debugProtocol.json

See the class comment on DebugMessageTypeAdapterFactory for details on
mapping between Debug Server Protocol and JSON-RPC 2.0.

Signed-off-by: Jonah Graham <jonah@kichwacoders.com>

eclipsewebmaster commented:
Issue tracker reference:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=130

mickaelistria commented:
@svenefftinge @spoenemann @akosyakov Is there anything I can do to help on this topic? I'm eager to see progress there and can allocate most of my time.

svenefftinge commented:
@jonahkichwacoders thanks for the PR. 
Could we vote you in as a committer and maintainer of this part?

spoenemann commented on ide/build.gradle:
```diff
@@ -30,6 +38,7 @@ oomphIde {
 	addProjectFolder('../org.eclipse.lsp4j')
 	addProjectFolder('../org.eclipse.lsp4j.generator')
 	addProjectFolder('../org.eclipse.lsp4j.jsonrpc')
+	addProjectFolder('../org.eclipse.dsp4j.jsonrpc')
```
I'm not sure about the namespace `org.eclipse.dsp4j`, since theoretically that could cause name clashes with other Eclipse projects. On the other hand, I like the symmetry with `org.eclipse.lsp4j`. What do others think about this?

An alternative project name could be `org.eclipse.lsp4j.jsonrpc.debug`

jonahkichwacoders replied:
OK, depending on decision here will affect name of the interfaces containing plug-in. I was originally planning on org.eclipse.dsp4j, but perhaps org.eclipse.lsp4j.debug makes more sense.

svenefftinge replied:
Yes, IIRC there even is some rule at eclipse for java packages to use the project name as prefix.

spoenemann commented on org.eclipse.dsp4j.jsonrpc/src/main/java/org/eclipse/dsp4j/jsonrpc/adapters/DebugMessageTypeAdapterFactory.java:
```diff
@@ -0,0 +1,493 @@
+			return result;
+		}
+
+		protected Object parseParams(JsonReader in, String method) throws IOException {
```
We should make `MessageTypeAdapterFactory` extensible in order to avoid duplicating code here. I'll prepare a PR for that.

jonahkichwacoders replied:
OK, there is certainly some duplication between the two.

spoenemann replied:
Merged #135.

jonahkichwacoders replied:
I updated the debug code to to use this new code. I also slightly refactored the `MessageTypeAdapterFactory` to access some of its logic from the subclass.

spoenemann commented on org.eclipse.dsp4j.jsonrpc/src/main/java/org/eclipse/dsp4j/jsonrpc/json/DebugMessageJsonHandler.java:
```diff
@@ -0,0 +1,38 @@
+		super(supportedMethods, configureGson);
+	}
+
+	public GsonBuilder getDefaultGsonBuilder() {
```
This additional class could be avoided by passing a `Consumer<GsonBuilder>` to the default `MessageJsonHandler`. When multiple type adapter factories are configured for the same type, the last one should win.

jonahkichwacoders replied:
OK, the advantage of your approach too is that the other type factories don't need to be duplicated. A lot of the tests instantiate `DebugMessageJsonHandler` directly, so I'll probably want a method somewhere that constructs a `MessageJsonHandler` correctly for the tests in the same way as it is constructed for by the `DebugLauncher`. Perhaps a method in the `DebugLauncher` itself.

jonahkichwacoders replied:
I am having trouble writing a sensible method in DebugLauncher, at the moment it looks like the sample below, which seems uglier than having the extra class, ideas welcome.

The reason for the complication is the circular dependency between the `MessageJsonHandler` and the `DebugMessageTypeAdapterFactory`.

My next update will have the `DebugMessageJsonHandler` still, but rather than repeating everything in `getDefaultGsonBuilder` I'll call super's one and then add on the `DebugMessageTypeAdapterFactory`.

If you have a thought on something I overlooked, I would be pleased for the feedback.

```java

	static MessageJsonHandler createMessageHandler(Map<String, JsonRpcMethod> supportedMethods,
			Consumer<GsonBuilder> configureGson) {
		DebugMessageTypeAdapterFactory adapterFactory[] = new DebugMessageTypeAdapterFactory[] {null};
		Consumer<GsonBuilder> consumer = (builder) -> {
			builder.registerTypeAdapterFactory(adapterFactory[0]);
			configureGson.accept(builder);
		};
		MessageJsonHandler messageJsonHandler = new MessageJsonHandler(supportedMethods, consumer);
		adapterFactory[0] = new DebugMessageTypeAdapterFactory(messageJsonHandler);
		return messageJsonHandler;
	}
```

spoenemann replied:
Hmmm, we could change the `configureGson` parameter in MessageJsonHandler to `BiConsumer<GsonBuilder, MessageJsonHandler>`, but that's a breaking change. @akosyakov, @svenefftinge do you think this is worth it? Would help for cases when someone wants to customize parsing / serialization of messages.

jonahkichwacoders replied:
We could also add an additional constructor and deprecate the existing one. That would prevent it being a breaking change.

spoenemann replied:
Right, we don't even need to deprecate the old one.

jonahkichwacoders replied:
I have not made this change yet. The current version, and the new constructor we discussed suffers from a problem. It passes a reference to `this` from the constructor when the object is not fully created, so creating new API that required and actually enforced passing around an incompletely constructed object seems like a bad idea. If you would like me to pursue this further (e.g. add an `initialize()` method for example) I can raise and issue and work on that separately?

spoenemann commented on org.eclipse.dsp4j.jsonrpc/src/test/java/org/eclipse/dsp4j/jsonrpc/test/json/DebugMessageJsonHandlerTest.java:
```diff
@@ -0,0 +1,582 @@
+	}
+
+	@Test
+	public void testParamsParsing_AllOrders() {
```
This parsing test that tries all iterations would also be nice for the `MessageJsonHandlerTest` in `org.eclipse.lsp4j.jsonrpc`.

jonahkichwacoders replied:
OK, I will contribute it back to both places. I also realised on revisiting that the idea was to remove the _02/_03/etc versions of the methods, so I will do that too.

jonahkichwacoders replied:
I have added the `_AllOrders` method to `MessageJsonHandlerTest` and extended it to do all the types of messages.

spoenemann commented:
@eclipsewebmaster 

> Issue tracker reference:
> https://bugs.eclipse.org/bugs/show_bug.cgi?id=130

That makes no sense. Has probably been triggered by Jonahs mention of `Bug #130`. So better use the term "issue" in this context :-D

jonahkichwacoders commented:
> Could we vote you in as a committer and maintainer of this part?
Yes.

jonahkichwacoders commented:
>Thank you for this proposal! The code seems to me of very high quality. I would support your nomination as a committer :-)

Thanks.

> Are you planning to also implement service interfaces with corresponding data structures in the style of org.eclipse.lsp4j.services.LanguageServer? Or have you already done that?

I am planning to create a PR in the coming days with the equivalent for the debug protocol. It is generated from the debug protocol json schema. Originally I was going to put it in a plug-in called org.eclipse.dsp4j, but based on your other comments, perhaps org.eclipse.lsp4j.debug would make more sense.

jonahkichwacoders commented:
> So better use the term "issue" in this context :-D

OK.

spoenemann commented:
It seems like the word "bug" is not necessary for triggering the Bugzilla bot, see eclipse/kapua#131

jonahkichwacoders commented:
@spoenemann I have updated the code with all your feedback and it is again ready for review.

This update has the next part too, the service protocol files. As the DSP has a [Json Schema](https://github.com/jonahkichwacoders/vscode-debugadapter-node/blob/53e08384d5ab0786a053ae24acdcfa0a081fb95f/debugProtocol.json) that defines it, the `DebugProtocol.java`, `IDebugProtocolClient` and `IDebugProtocolServer` are auto-generated using a [generator](https://github.com/jonahkichwacoders/vscode-debugadapter-node/blob/53e08384d5ab0786a053ae24acdcfa0a081fb95f/src/generator.ts) derived from the [typescript generator](https://github.com/Microsoft/vscode-debugadapter-node/blob/99528cd3c2db13f3fd78047c5b6967bf13212c23/src/generator.ts) used by vscode. Please see the instructions for the generator in the root of the `org.eclipse.lsp4j.debug` plug-in. I don't know if the licensing of the generator allows it to live in the lsp4j code base. I also don't know if continuing to generate it automatically as the protocol evolves is the best plan either, so perhaps the generator is only useful to bootstrap the project?

spoenemann commented on org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.java:
```diff
@@ -0,0 +1,11631 @@
+	 * completed, by executing a debugger statement etc.
+	 * </p>
+	 */
+	public static class StoppedEventArguments {
```
The class structure should not be nested, but flat, in order to be consistent with the LSP implementation.

jonahkichwacoders replied:
Probably it would be better to generate a `DebugProtocol.xtend` with `@JsonRpcData` tags and that gets us even closer. It would be a 2 stage generation process (first the typescript one then the o.e.lsp4j.generator one), but then the `DebugProtocol.xtend` would be much readable (and shorter). WDYT?

spoenemann replied:
Two-stage generation sounds a bit scary at first. But a nice side effect is that all the derived members (getters, setters, equals, hashCode) are consistent in both protocols. And I like the idea to have a more readable version of the Java data structures.

spoenemann commented on org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.java:
```diff
@@ -0,0 +1,11631 @@
+		 * Possible values include - but not limited to those defined in {@link Reason}
+		 * </p>
+		 */
+		public String reason;
```
The fields should not be public, but accessible via getters, in order to be consistent with the LSP implementation.

spoenemann commented on org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/services/IDebugProtocolClient.java:
```diff
@@ -0,0 +1,218 @@
+	 * </ul>
+	 */
+	@JsonNotification
+	default void initialized() {
```
Why these default implementations? I'd remove them.

jonahkichwacoders commented on org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.java:
```diff
@@ -0,0 +1,11631 @@
+	/**
+	 * <p>
+	 * Version of debugProtocol.json this class was derived from.
+	 * </p>
```
I can add some additional logic to the generator to only add the `<p>` between paragraphs and omit them on single paragraph comments.

jonahkichwacoders commented:
@spoenemann Its ready, I have hopefully addressed all your concerns adequately. With the generation of `DebugProtocol.xtend` and using the existing generator as the second-pass generator makes it much easier to read `DebugProtocol.xtend` and the generated javadoc still looks fine.

spoenemann commented on org.eclipse.lsp4j.debug/src/main/java/org/eclipse/lsp4j/debug/DebugProtocol.xtend:
```diff
@@ -0,0 +1,2470 @@
+ * <p>
+ * Auto-generated from debugProtocol.json schema version 1.24.0. Do not edit manually.
+ */
+class DebugProtcol {
```
I think this class could be removed now.
